monitoring/google/cloud/monitoring/_dataframe.py

```diff
@@ -0,0 +1,143 @@
+
+TOP_RESOURCE_LABELS = (
+    'project_id',
+    'aws_account',
+    'location',
+    'region',
+    'zone',
+)
+
+
+def _extract_header(time_series):
+    """Return a copy of time_series with the points removed."""
+    return TimeSeries(
+        metric=time_series.metric,
+        resource=time_series.resource,
+        metric_kind=time_series.metric_kind,
+        value_type=time_series.value_type
+    )
+
+
+def _extract_labels(time_series):
+    """Build the combined resource and metric labels, with resource_type."""
+    labels = {'resource_type': time_series.resource.type}
+    labels.update(time_series.resource.labels)
+    labels.update(time_series.metric.labels)
+    return labels
+
+
+def _extract_value(typed_value):
+     """Extract the value from a TypedValue."""
+     value_type = typed_value.WhichOneof('value')
+     return typed_value.__getattribute__(value_type)
+
+
+def _build_dataframe(time_series_iterable,
+                     label=None, labels=None):  # pragma: NO COVER
+    """Build a :mod:`pandas` dataframe out of time series.
+
+    :type time_series_iterable:
+        iterable over :class:`~google.cloud.monitoring_v3.types.TimeSeries`
+    :param time_series_iterable:
+        An iterable (e.g., a query object) yielding time series.
+
+    :type label: str
+    :param label:
+        (Optional) The label name to use for the dataframe header. This can be
+        the name of a resource label or metric label (e.g.,
+        ``"instance_name"``), or the string ``"resource_type"``.
+
+    :type labels: list of strings, or None
+    :param labels:
+        A list or tuple of label names to use for the dataframe header.
+        If more than one label name is provided, the resulting dataframe
+        will have a multi-level column header.
+
+        Specifying neither ``label`` or ``labels`` results in a dataframe
+        with a multi-level column header including the resource type and
+        all available resource and metric labels.
+
+        Specifying both ``label`` and ``labels`` is an error.
+
+    :rtype: :class:`pandas.DataFrame`
+    :returns: A dataframe where each column represents one time series.
+    """
+    import pandas   # pylint: disable=import-error
+
+    if labels is not None:
+        if label is not None:
+            raise ValueError('Cannot specify both "label" and "labels".')
+        elif not labels:
+            raise ValueError('"labels" must be non-empty or None.')
+
+    columns = []
+    headers = []
+    for time_series in time_series_iterable:
+        pandas_series = pandas.Series(
+            data=[_extract_value(point.value) for point in time_series.points],
+            index=[point.interval.end_time.ToNanoseconds()
+                   for point in time_series.points],
+        )
+        columns.append(pandas_series)
+        headers.append(_extract_header(time_series))
+
+    # Implement a smart default of using all available labels.
+    if label is None and labels is None:
```

Comment:
If you made the change above, this test would change to `if labels is not None:`.

Also, do you want `is None` or just falsity? Do you want an empty sequence to behave differently from `None`? (An alternate way to ask that: is having zero loop iterations in 122-124 valid?)

Reply:
Check for falsity is fine. None or [] would behave the same now. Earlier an error was being thrown when labels == [].

In 122-124, we will always have 1 or more iterations. When both label and labels are not passed in, we return all the available "labels" or headers in the data.